### PR TITLE
Generic socket for both secure and non secure tcp socket

### PR DESCRIPTION
**Pull request by mchindri**
- Add generic implementation
- Fix structure name in wifi
- Make mqtt use generic interface
- Refactor tcp_socket to fit the new interface
- Create a template for secure tcp that fits the generic socket

### CODE REVIEW

File: include/socket.h [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+/******************************************************************************/
+/***************************** Include Files **********************************/
+/******************************************************************************/
+
+#include <stdint.h>
+#include "network_interface.h"
+
+/**
+ *
+ */
```
Comment (adrimbarean) [inline include/socket.h:53]: nitpick: empty comment

File: network/data_link/wifi.c [hunk 2]
```diff
@@ -166,7 +166,7 @@ int32_t wifi_init(struct wifi_desc **desc, struct wifi_init_param *param)
 
 	wifi_init_intarface(ldesc);
 
-	result = at_run_cmd(ldesc->at, AT_RESET, AT_EXECUTE_OP, NULL);
+	//result = at_run_cmd(ldesc->at, AT_RESET, AT_EXECUTE_OP, NULL);
```
Comment (adrimbarean) [inline network/data_link/wifi.c:169]: Commented code should be removed if it's not needed.

File: include/socket.h (hunk 1 above)
Comment (mchindri) [reply include/socket.h]: solved

File: network/data_link/wifi.c (hunk 2 above)
Comment (mchindri) [reply network/data_link/wifi.c]: this change is removed. (was commented only for faster testing)

**Comment by mchindri** [comment]
v2
- remove last commit with secure tcp template (the implementation is at https://github.com/analogdevicesinc/no-OS/pull/586)
- fix astyle
- move socket_interface.sock to sock_desc.sock
- remove empty comment from socket.h
- move comments from tcp_socket.c to socket.c
- undo change of commenting at_run_command

**Comment by mchindri** [comment]
v3:
 - fix astyle

File: include/socket.h [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+ * @brief Socket descriptor
+ */
+struct socket_desc;
+
+/**
+ * @struct tcp_socket_init_param
+ * @brief Parameter to initialize a TCP Socket
+ */
+struct tcp_socket_init_param {
+	//No parameters
```
Comment (mirceacaprioru) [inline include/socket.h:69]: will this be populated in a future commit ?

File: include/socket.h [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+struct socket_init_param {
+	/** Network interface. Ex: Wifi, Ethernet */
+	struct network_interface	*net;
+	/** Reference to one of the values */
+	enum socket_type		type;
+	/** Socket init param */
+	union {
+		struct tcp_socket_init_param	*tcp;
+		struct stcp_socket_init_param	*secure_tcp;
+	}				init_param;
```
Comment (mirceacaprioru) [inline include/socket.h:101]: looks like inconsistent indentation

File: include/socket.h [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+ * @brief Parameter to initialize a Socket
+ */
+struct socket_init_param {
+	/** Network interface. Ex: Wifi, Ethernet */
+	struct network_interface	*net;
+	/** Reference to one of the values */
+	enum socket_type		type;
+	/** Socket init param */
+	union {
+		struct tcp_socket_init_param	*tcp;
```
Comment (mirceacaprioru) [inline include/socket.h:99]: this one does not have any members, so a union like this looks a wierd.

File: network/data_link/wifi.c [hunk 6]
```diff
@@ -59,7 +59,7 @@
 /******************************************************************************/
```
Comment (mirceacaprioru) [inline network/data_link/wifi.c:59]: typo in commit message: Raname -> Rename

File: network/data_link/wifi.c [hunk 7]
```diff
@@ -59,7 +59,7 @@
...
 
 /* Structure storing data used by a socket */
-struct socket_desc {
+struct wifi_sock {
```
Comment (mirceacaprioru) [inline network/data_link/wifi.c:62]: the name here is not consistent with the commit title. Should it be wifi_desc ?

File: network/transport/tcp_socket.c [hunk 8]
```diff
@@ -44,20 +44,58 @@
...
+			       uint32_t len);
+
+/******************************************************************************/
+/********************** Macros and Constants Definitions **********************/
+/******************************************************************************/
+
+/* Interface to used by generic socket.c */
+const struct socket_interface tcp_socket_interface = {
+	.protocol = PROTOCOL_TCP,
+	.init = (int32_t (*)())tcp_socket_init,
```
Comment (mirceacaprioru) [inline network/transport/tcp_socket.c:89]: no need for the cast.

File: network/transport/tcp_socket.c [hunk 9]
```diff
@@ -70,20 +108,23 @@ struct tcp_socket_desc {
...
-		    struct tcp_socket_init_param *param)
+static int32_t tcp_socket_init(struct tcp_socket_desc **desc,
+			       struct network_interface *net,
+			       struct tcp_socket_init_param *param)
 {
 	struct tcp_socket_desc	*ldesc;
 	int32_t			ret;
 
-	if (!desc || !param)
+	UNUSED_PARAM(param);
```
Comment (mirceacaprioru) [inline network/transport/tcp_socket.c:111]: i think this is unused and also it is an empty structure. It looks redundant to have it like this. Maybe just to have a socket_init_param for both tcp and tls would make more sense and just not use it for tcp if you do not need anything from it. To have an empty structure is useless.

File: network/transport/tcp_socket.c [hunk 10]
```diff
@@ -149,10 +190,12 @@ int32_t socket_send(struct tcp_socket_desc *desc, const void *data,
...
-int32_t socket_recv(struct tcp_socket_desc *desc, void *data, uint32_t len)
+static int32_t tcp_socket_recv(struct tcp_socket_desc *desc, void *data,
+			       uint32_t len)
 {
 	if (!desc)
 		return FAILURE;
 
 	return desc->net->socket_recv(desc->net->net, desc->id, data, len);
 }
+
```
Comment (mirceacaprioru) [inline network/transport/tcp_socket.c:188]: nitpick: extra line

File: include/socket.h [hunk 11]
```diff
@@ -0,0 +1,134 @@
+/***************************************************************************//**
```
Comment (adrimbarean) [inline include/socket.h:1]: Does this file make [tcp_socket.h](https://github.com/analogdevicesinc/no-OS/blob/master/include/tcp_socket.h) redundant?

Comment (mchindri) [reply include/socket.h:1]: Yes, tcp_socket.h is removed in one commit

File: include/socket.h (hunk 4 above)
Comment (mchindri) [reply include/socket.h:101]: init_parma is indented with type, net and so on

File: include/socket.h (hunk 5 above)
Comment (mchindri) [reply include/socket.h:99]:
```
        //socket.c
 	switch (param->type) {
	case TCP_SOCKET:
		init_param = (void *)param->init_param.tcp;
		break;
	case SECURE_TCP_SOCKET:
		init_param = (void *)param->init_param.secure_tcp;
		break;
}
```
This is how it is used. With the type you choose which init param to use.
For the type TCP_SOCKET no init param is needed. I created it just for consistency.
For example if new a type is added (ex. UDP_SOCKET), new param should be added too.

File: network/transport/tcp_socket.c (hunk 8 above)
Comment (mchindri) [reply network/transport/tcp_socket.c:89]: It is needed to avoid a warning

**Comment by mchindri** [comment]
v4
 - Fix wifi commit message
 - Remove extra line

File: network/transport/tcp_socket.c (hunk 10 above)
Comment (mchindri) [reply network/transport/tcp_socket.c]: removed

File: network/transport/tcp_socket.c (hunk 9 above)
Comment (mchindri) [reply network/transport/tcp_socket.c:111]: Explained here https://github.com/analogdevicesinc/no-OS/pull/581#discussion_r423653253

File: network/data_link/wifi.c (hunk 7 above)
Comment (mchindri) [reply network/data_link/wifi.c:62]: Changed in commit message

File: network/data_link/wifi.c (hunk 6 above)
Comment (mchindri) [reply network/data_link/wifi.c:59]: solved

File: include/socket.h (hunk 3 above)
Comment (mchindri) [reply include/socket.h:69]: https://github.com/analogdevicesinc/no-OS/pull/581#discussion_r423653253

**Comment by adrimbarean** [comment]
Last commit needs signature.

**Comment by mchindri** [comment]
v5:
 - Sign last commit

**Pull request closed without merging**